sensor/kubernetes/eventpipeline/output/output_impl.go

```diff
@@ -0,0 +1,70 @@
+package output
+
+import (
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/sensor/common/detector"
+	"github.com/stackrox/rox/sensor/kubernetes/eventpipeline/component"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+type outputQueueImpl struct {
+	innerQueue   chan *component.ResourceEvent
+	stopSig      *concurrency.Signal
+	forwardQueue chan *central.MsgFromSensor
+	detector     detector.Detector
+}
+
+// Send sends a ResourceEvent message to the inner queue
+func (q *outputQueueImpl) Send(msg *component.ResourceEvent) {
+	q.innerQueue <- msg
+}
+
+// ResponsesC returns the MsgFromSensor channel
+func (q *outputQueueImpl) ResponsesC() <-chan *central.MsgFromSensor {
+	return q.forwardQueue
+}
+
+// Start starts the outputQueueImpl component
+func (q *outputQueueImpl) Start() error {
+	go q.runOutputQueue()
+	return nil
+}
+
+// Stop stops the outputQueueImpl component
+func (q *outputQueueImpl) Stop(_ error) {
+	defer close(q.innerQueue)
+	defer close(q.forwardQueue)
+	q.stopSig.Signal()
```

Comment:
I think there is some odd behavior here. `q.stopSig` is a pointer passed in the constructor. It makes sense to use it if `output` fails unexpectedly and we want to propagate the error up. However, calling it when it gets closed conflicts with the parent `Close()` call. Which does the same thing:

```go
// Stop implements common.SensorComponent
func (p *eventPipeline) Stop(err error) {
	defer close(p.eventsC)
	for _, c := range p.components {
		c.Stop(err)
	}
	p.stopSig.Signal()
}
```

This is confusing because `stopSig.Signal()` will be called more than once (pipeline and output) and if the behavior is copied, it will be called in other sub-components as well.

Reply:
That was a mistake on my part. The idea is to have independent signals for each component. Similar to how it’s done for `SensorComponents`. This way each component is in charge of stopping itself. Now I see your point of wanting to propagate an error in a sub-component up. I think if we want to do that we should use a different signal (e.g. `errorSignal`) and have the parent component listen to the error signal in its main loop. WDYT?